testsuite/tests/quotation/quotation_typing.ml

```diff
@@ -312,3 +312,214 @@ let eta (type a) (x : a expr) : a expr = <[ $x ]>
 [%%expect {|
 val eta : 'a expr -> 'a expr = <fun>
 |}]
+
+(* These should fail kinding checks *)
+
+let id : 'a -> 'a = fun x -> x
+
+[%%expect{|
+val id : 'a -> 'a = <fun>
+|}]
+
+(* But this should preferably work at some point *)
+let id' : <[<['a]>]> -> <[<['a]>]> = id
+
+[%%expect{|
+Line 1:
+Error: Values do not match:
+         val id' : ('a : any). 'a -> 'a
+       is not included in
+         val id' : ('a : any). 'a -> 'a
+       The type "$('a) -> $('a)" is not compatible with the type "'a -> 'a"
+       Type "$('a)" = "$('a)" is not compatible with type "'a" = "'a"
+       Type "'a" is not compatible with type "<['a]>"
+|}]
+
+let foo : <[<['a]>]> -> <['a]> -> 'a -> 'a = fun _ _ a -> a
+
+[%%expect{|
+Line 1, characters 26-28:
+1 | let foo : <[<['a]>]> -> <['a]> -> 'a -> 'a = fun _ _ a -> a
+                              ^^
+Error: Type variable "'a" is used inside a quotation (<[ ... ]>),
+       it already occurs inside 2 layers of quotation (<[ ... ]>).
+       Hint: Consider using "<['a]>".
+|}]
+
+(* Inclusion checks *)
+
+(*  [M1] and [M2] pass due to stage normalisation *)
```

Comment:
We don't use the term "stage normalisation" anywhere in the code, so it would be good to elaborate what we mean by this.